### PR TITLE
[c++] Use extern template for common code

### PR DESCRIPTION
**Pull request by ara-ayvazyan**

### CODE REVIEW

File: cpp/inc/bond/core/detail/extern_macro.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+
+#define BOND_DETAIL_APPLY_FUNC(r, p) \
+    BOOST_PP_EXPAND(BOOST_PP_SEQ_HEAD(p) BOOST_PP_EXPAND(BOOST_PP_SEQ_TO_TUPLE(BOOST_PP_SEQ_POP_FRONT(p))))
+
+#define BOND_DETAIL_PRECOMPILE(F, S) \
+    BOOST_PP_SEQ_FOR_EACH_PRODUCT(BOND_DETAIL_APPLY_FUNC, ((F))S)
+
+#define BOND_DETAIL_BUILTIN_READERS \
+    (CompactBinaryReader<InputBuffer>) \
```
Comment (ara-ayvazyan) [inline cpp/inc/bond/core/detail/extern_macro.h]: Here we should probably take into account the protocol enable/disable macros.

File: cpp/inc/bond/core/detail/extern_macro.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+#pragma once
+
+#include <bond/core/config.h>
+
+#if !defined(BOND_COMPACT_BINARY_PROTOCOL) \
+ && !defined(BOND_SIMPLE_BINARY_PROTOCOL) \
+ && !defined(BOND_FAST_BINARY_PROTOCOL) \
+ && !defined(BOND_SIMPLE_JSON_PROTOCOL)
+#error None of the built-in protocols is enabled
```
Comment (chwarr) [inline cpp/inc/bond/core/detail/extern_macro.h]: What about a message like this?

"None of the built-in protocols are enabled. Either enable one of the built-in protocols or set BOND_LIB_TYPE to BOND_LIB_TYPE_HEADER."

File: cpp/inc/bond/core/detail/extern_macro.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+#endif
+
+#include <boost/preprocessor/seq/seq.hpp>
```
Comment (chwarr) [inline cpp/inc/bond/core/detail/extern_macro.h]: Minor: alpha sort if possible. If not, add a comment.

File: cpp/inc/bond/core/detail/parser_extern.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include "extern_macro.h"
```
Comment (chwarr) [inline cpp/inc/bond/core/detail/parser_extern.h]: Minor: missing config.h

File: cpp/inc/bond/core/detail/select_protocol_extern.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include "extern_macro.h"
```
Comment (chwarr) [inline cpp/inc/bond/core/detail/select_protocol_extern.h]: Minor: missing config.h

File: cpp/inc/bond/core/detail/typeid_value_extern.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include "extern_macro.h"
```
Comment (chwarr) [inline cpp/inc/bond/core/detail/typeid_value_extern.h]: Minor: missing config.h

File: cpp/inc/bond/core/detail/value_extern.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include "extern_macro.h"
```
Comment (chwarr) [inline cpp/inc/bond/core/detail/value_extern.h]: Minor: missing config.h

File: cpp/src/bond/core/detail/typeid_value.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <bond/core/bond.h>
```
Comment (chwarr) [inline cpp/src/bond/core/detail/typeid_value.cpp]: Minor: missing config.h

File: cpp/src/bond/core/parser.cpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <bond/core/bond.h>
```
Comment (chwarr) [inline cpp/src/bond/core/parser.cpp]: Minor: missing config.h

File: cpp/src/bond/core/select_protocol.cpp [hunk 10]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <bond/core/bond.h>
```
Comment (chwarr) [inline cpp/src/bond/core/select_protocol.cpp]: Minor: missing config.h

**Pull request merged**